Delegate tree building to the parser

little-bobby-tables opened the pull request:
Removing the intermediate representation reduces the complexity of the parser and makes it easier to perform line-independent transforms (handling `if-else`, for instance).

I had to add trailing whitespace removal to the preprocessor because I couldn't handle certain edge cases like "blank lines after if-else" (*test/rendering_test.ex*) without severely complicating the grammar.

little-bobby-tables committed:
add new node types

little-bobby-tables committed:
transform input into a tree

little-bobby-tables committed:
remove trailing whitespace in preprocessor

little-bobby-tables committed:
modify parsing helpers to work with the new transforms

little-bobby-tables committed:
rewrite the compiler to render new node types

little-bobby-tables committed:
remove Slime.Tree

little-bobby-tables committed:
clean up consecutive blank lines

doomspork commented:
Thank you @little-bobby-tables!  @Rakoth and/or myself will take a peek at this and let you know if we have any feedback 😁

Rakoth commented on lib/slime/parser/transform.ex:
```diff
@@ -190,50 +209,28 @@ defmodule Slime.Parser.Transform do
+      "/" -> {[], true}
+      "" -> {[], false}
+      [] -> {[], false}
+      (%EExNode{} = eex) -> {[eex], false}
```
We could use `content` variable here and in next match:
```
%EExNode{} -> {[content], false}
```

Rakoth commented on lib/slime/compiler.ex:
```diff
@@ -3,41 +3,77 @@ defmodule Slime.Compiler do
+    body = cond do
+      tag.closed             -> "<" <> tag_head <> "/>"
+      name in @void_elements -> "<" <> tag_head <> ">"
+      :otherwise             -> "<" <> tag_head <> ">"
```
matter of taste, but I would prefer line-break after `->` in this case

Rakoth commented on lib/slime/compiler.ex:
```diff
@@ -3,41 +3,77 @@ defmodule Slime.Compiler do
-    |> Enum.map(fn branch -> render_branch(branch) end)
-    |> Enum.join(lines_sep)
+  def compile([]), do: ""
+  def compile([h | tree]) do
```
guard will do better here `compile(tags) when is_list(tags)`

Rakoth commented on src/slime_parser.peg.eex:
```diff
@@ -1,32 +1,33 @@
 
-% NOTE: Leading whitespace is handled separately for comments and verbatim text
-tag <- comment / verbatim_text / space? tag_item / blank:(space? &eol);
+tags <- (tag crlf*)+;
```
Should we order rules as they appear in grammar, from top to bottom for easier reading?

Rakoth commented on src/slime_parser.peg.eex:
```diff
@@ -1,32 +1,33 @@
+nested_tags <- crlf+ indent tags dedent;
 
-tag_item <- (embedded_engine / inline_html / code / html_tag);
+inline_html <- &'<' text_content nested_tags?;
```
Am I missing something, or html with nested children should always have a closing inline html part, so the following examples should raise a parsing error.
```
<img />
  | test
```
```
<div>
  | test
</div>
  | test 1
```

little-bobby-tables replied:
Slim does not check for the closing tag. It's an interesting idea — it could be handled similar to `if-else` — but I think it is out of scope of this PR, yes.

Rakoth commented on src/slime_parser.peg.eex:
```diff
@@ -1,32 +1,33 @@
-code <- inline:('=' '='? tag_spaces? / '-') space? code:code_lines;
-code_lines <- code_line / code_line_with_brake crlf (indent code_lines dedent / code_lines);
+simple_tag <- tag:tag_shortcut spaces:tag_spaces space?
+  content:attributes_with_content children:nested_tags?;
```
There are some invalid test cases, currently supported by accident. In ruby-slim tag with inline content could not have nested children tags, (if there are nested lines, they treated as inline content continuation). In some cases it as already not handled by slime, so I think it is time to fix this behaviour, or at least have a parsing error for this case.
For example:
```
div test
  test
```
in slim renders to
```
<div>test
test</div>
```

Rakoth reviewed (commented):
This is a great work done, @little-bobby-tables ! I left some comments, there are two about parser grammar, which we could open as separate issues, if you think it is not in scope of this changes.

Rakoth commented:
There is also some problem with code continuation lines:
```
= Enum.join([1,
  2])

p test
```
raises a parser error about unexpected dedent on line 3

little-bobby-tables committed:
minor readability improvements

little-bobby-tables committed:
tags with inline content shouldn't have nested children

little-bobby-tables committed:
unify simple_tag and inline_tag

Inline tags in Slim can have attributes, both normal and wrapped.

little-bobby-tables committed:
parse inline tag text

little-bobby-tables committed:
fix integration tests

little-bobby-tables committed:
skip indents in code lines broken by ,

little-bobby-tables committed:
take advantage of the preprocessor stripping trailing whitespace

little-bobby-tables commented:
@Rakoth, I think I've addressed the issues you've brought up. I couldn't pinpoint exactly why the code continuation rule was failing; I believe it was due to some other rule that used to consume dedents past it, but I'm not sure. Either way, since the preprocessor already skips indentation for lines broken by a backslash, I've just added commas to the regular expression there and that seemed to solve it.

I also haven't replied to your comment about grammar reordering. I don't feel strongly about it, it's probably better to do it outside of this changeset.

@doomspork: the PR now includes a breaking change concerning the behavior of nodes with inline content. What used to be rendered as nested tags:
```
h1 Secret:
  = word
```
is now rendered as text, so as to fit this example from the Slim documentation:
```
p No items found. Please add some inventory.
  Thank you!
```
WDYT?

Rakoth reviewed (approved):
👍 LGTM

Rakoth commented:
@doomspork I think we good to merge this PR, any objections?

doomspork commented:
No sir @Rakoth!  This looks good to me.

Thank you again @little-bobby-tables, this is great!
